### PR TITLE
Fixed issues with moving platforms

### PR DESCRIPTION
**Pull request by mbernard2**
Addresses  #401 and  #1389.

When an object is on top of a moving static object (e.g. platform) or a tilemap, every movement made by the static object (or tilemap) is also applied to all objects standing on top of it.

Also, easing works much better.

Here are a few examples:
- Normal back-and-forth:
  
- Trampolines now work on downwards moving platforms:
  
- Player can stand on top of a block which itself is on top of a platform. Here's an example with unisolid tiles:
  
- Unisolid slopes:

### CODE REVIEW

**Comment by weluvgoatz** [comment]
Wasn't "slightly above ground" there for jump buffering?? Jump buffering is a really great feature and the game would play much much worse without it.

**Comment by mbernard2** [comment]
> Wasn't "slightly above ground" there for jump buffering?? Jump buffering is a really great feature and the game would play much much worse without it.

Does jump buffering have to do with being able to press "jump" while still in the air, and jump when on the ground ?

If so, I just tested it, and it still works.

**Comment by weluvgoatz** [comment]
I tested it as well, and it works properly still.  Thank you, just making sure :-)

**Comment by Semphriss** [comment]
Yay, finally! :D

**Comment by mbernard2** [comment]
Don't merge yet, I found a bug that only occurs on this branch (which means it's a regression).

Basically, objects stacked on top of each other slowly move into the ground and eventually fall off the screen.

File: src/collision/collision_system.hpp [hunk 1]
```diff
@@ -45,6 +47,11 @@ class CollisionSystem final
       case (or not). */
   void update();
 
+  const std::shared_ptr<CollisionGroundMovementManager>& get_ground_movement_manager()
```
Comment (tobbi) [inline src/collision/collision_system.hpp:50]: I think this method can be marked `const`

**Comment by weluvgoatz** [comment]
Any intention to still work on this? It would be nice if it was implemented, once the bug(s) were resolved at least.

**Comment by mbernard2** [comment]
Yes.

**Comment by mbernard2** [comment]
Almost done!

**Comment by mbernard2** [comment]
FINALLY!
Objects can now be stacked indefinitely over one another, on moving platforms with or without easing or on solid ground. Tested with trampolines, unisolids, platform objects, moving tilemaps...

**Comment by tobbi** [comment]
Thank you very much for this. This issue has bugged us for such a long time. I will review your code this evening and then merge it once all is good. :)

**Comment by Semphriss** [comment]
Hurray!

Thanks so much for your contribution!

**Comment by weluvgoatz** [comment]
There seems to be a bug where Tux can no longer ride on sideways-facing ice crushers anymore, as he just slips off of them weirdly.

**Comment by mbernard2** [comment]
> There seems to be a bug where Tux can no longer ride on sideways-facing ice crushers anymore, as he just slips off of them weirdly.

As of f67974b, he can ride on "recovering" crushers but not on "crushing" crushers (i.e. in "crushing" mode). I'll fix this.

***

BTW can someone help me diagnose why the "bsd-solaris" check failed ?

**Comment by tobbi** [comment]
> BTW can someone help me diagnose why the "bsd-solaris" check failed ?

Seems to be an intermittent failure. I've re-triggered the builds.

**Comment by tobbi** [comment]
Thank you very much for this 🎉 I'm merging it.

🥳

**Pull request merged**